zebrad/tests/common/get_block_template_rpcs.rs

```diff
@@ -0,0 +1,164 @@
+use super::*;
+
+pub(crate) mod submit_block {
+
+    use std::path::PathBuf;
+
+    use color_eyre::eyre::{eyre, Context, Result};
+
+    use futures::TryFutureExt;
+    use indexmap::IndexSet;
+    use reqwest::Client;
+    use tower::{Service, ServiceExt};
+    use zebra_chain::{block::Height, parameters::Network, serialization::ZcashSerialize};
+    use zebra_state::HashOrHeight;
+    use zebra_test::args;
+
+    use crate::common::{
+        cached_state::{copy_state_directory, start_state_service_with_cache_dir},
+        config::{persistent_test_config, testdir},
+        launch::ZebradTestDirExt,
+        lightwalletd::random_known_rpc_port_config,
+    };
+
+    use super::cached_state::{load_tip_height_from_state_directory, ZEBRA_CACHED_STATE_DIR};
+
+    async fn get_future_block_hex_data(
+        network: Network,
+        zebrad_state_path: &PathBuf,
+    ) -> Result<Option<String>> {
+        tracing::info!(
+            ?zebrad_state_path,
+            "getting cached sync height from ZEBRA_CACHED_STATE_DIR path"
+        );
+
+        let cached_sync_height =
+            load_tip_height_from_state_directory(network, zebrad_state_path.as_ref()).await?;
+
+        let future_block_height = Height(cached_sync_height.0 + 1);
+
+        tracing::info!(
+            ?cached_sync_height,
+            ?future_block_height,
+            "got cached sync height, copying state dir to tempdir"
+        );
+
+        let copied_state_path = copy_state_directory(network, &zebrad_state_path).await?;
+
+        let mut config = persistent_test_config()?;
+        config.state.debug_stop_at_height = Some(future_block_height.0);
+
+        let mut child = copied_state_path
+            .with_config(&mut config)?
+            .spawn_child(args!["start"])?
+            .bypass_test_capture(true);
+
+        while child.is_running() {
+            tokio::task::yield_now().await;
+        }
+
+        let _ = child.kill(true);
```

Comment:
This seems to be a copy of the first part of `load_transactions_from_future_blocks()`. We could split that function instead of duplicating most of its code.

This code is missing some coverage:
* if we don't sync to the tip, the test won't cover the latest blocks, which is particularly important around network upgrades

And some error handling code:
* there is no timeout on the `zebrad` child task
* `sync_until()` checks Zebra's logs for errors, but `while child.is_running() { tokio::task::yield_now().await }` will keep running even if Zebra logs a failure

These things are already handled by `load_transactions_from_future_blocks()`.

Reply:
Noted. In hindsight I really should've renamed `LightwalletdTestType` and added a variant to it from the outset. I'll make sure to reference and address all of these comments when doing that.